coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala

```diff
@@ -0,0 +1,150 @@
+package filodb.coordinator.queryplanner
+
+import com.typesafe.config.ConfigFactory
+
+import filodb.coordinator.queryplanner.LogicalPlanUtils._
+import filodb.core.metadata.Dataset
+import filodb.core.query.{PromQlQueryParams, QueryContext}
+import filodb.query.{BinaryJoin, LabelValues, LogicalPlan, PeriodicSeriesPlan, SeriesKeysByFilters}
+import filodb.query.exec._
+
+case class PartitionAssignment(partitionName: String, endPoint: String, timeRange: TimeRange)
+
+trait PartitionLocationProvider {
+
+  def getPartitions(routingKey: Map[String, String], timeRange: TimeRange): List[PartitionAssignment]
+  def getAuthorizedPartitions(timeRange: TimeRange): List[PartitionAssignment]
+}
+
+class MultiPartitionPlanner(partitionLocationProvider: PartitionLocationProvider,
+                            localPlanner: QueryPlanner,
+                            localPartition: String,
+                            dataset: Dataset) extends QueryPlanner {
+
+  override def materialize(logicalPlan: LogicalPlan, qContext: QueryContext): ExecPlan = {
+
+    val tsdbQueryParams = qContext.origQueryParams
+
+    if(!tsdbQueryParams.isInstanceOf[PromQlQueryParams] || // We don't know the promql issued (unusual)
+      (tsdbQueryParams.isInstanceOf[PromQlQueryParams]
+        && !tsdbQueryParams.asInstanceOf[PromQlQueryParams].processRouting)) // Query was part of routing
+      localPlanner.materialize(logicalPlan, qContext)
+
+    else logicalPlan match {
+      case lp: BinaryJoin          => materializeBinaryJoin(lp, qContext)
+      case lp: LabelValues         => materializeLabelValues(lp, qContext)
+      case lp: SeriesKeysByFilters => materializeSeriesKeysFilters(lp, qContext)
+      case _                       => materializeSimpleQuery(logicalPlan, qContext)
+
+    }
+  }
+
+  private def getRoutingKeys(logicalPlan: LogicalPlan) = dataset.options.nonMetricShardColumns
+    .map(x => (x, LogicalPlan.getLabelValueFromLogicalPlan(logicalPlan, x)))
+
+  private def generateRemoteExec(queryContext: QueryContext, startMs: Long,
+                                 endMs: Long, endPoint: String): PromQlExec = {
+    val queryParams = queryContext.origQueryParams.asInstanceOf[PromQlQueryParams]
+    val promQlParams = PromQlQueryParams(ConfigFactory.parseString(s"""endpoint = "${endPoint}""""),
+      queryParams.promQl, startMs / 1000, queryParams.stepSecs, endMs / 1000, queryParams.spread, processFailure = true,
+      processRouting = false)
+    PromQlExec(queryContext, InProcessPlanDispatcher, dataset.ref, promQlParams)
```

Comment:
As mentioned offline, we may need a different ExecPlan for different Prom endpoints. The result schema would also be different for each.